src/storage/index/hnsw_index.cpp

```diff
@@ -0,0 +1,364 @@
+namespace storage {
+
+// NOLINTNEXTLINE(readability-make-member-function-const): Semantically non-const function.
+void HNSWIndexPartitionerSharedState::setTables(NodeTable* nodeTable, RelTable* relTable) {
+    lowerPartitionerSharedState->srcNodeTable = nodeTable;
+    lowerPartitionerSharedState->dstNodeTable = nodeTable;
+    lowerPartitionerSharedState->relTable = relTable;
+    upperPartitionerSharedState->srcNodeTable = nodeTable;
+    upperPartitionerSharedState->dstNodeTable = nodeTable;
+    upperPartitionerSharedState->relTable = relTable;
+}
+
+InMemHNSWLayer::InMemHNSWLayer(MemoryManager* mm, InMemHNSWLayerInfo info) : info{info} {
+    graph = std::make_unique<InMemHNSWGraph>(mm, info.numNodes, info.maxDegree);
+}
+
+void InMemHNSWLayer::insert(transaction::Transaction* transaction, common::offset_t offset,
+    common::offset_t entryPoint_, VisitedState& visited) {
+    if (entryPoint_ == common::INVALID_OFFSET) {
+        setEntryPoint(offset);
+        // The layer is empty. No edges need to be created.
+        return;
+    }
+    const auto closest = searchKNN(transaction, info.embeddings->getEmbedding(offset),
+        getEntryPoint(), info.maxConnectivity, info.efc, visited);
+    for (const auto& n : closest) {
+        insertRel(transaction, offset, n.nodeOffset);
+        insertRel(transaction, n.nodeOffset, offset);
+    }
+}
+
+common::offset_t InMemHNSWLayer::searchNN(transaction::Transaction* transaction,
+    common::offset_t node, common::offset_t entryNode) {
+    auto currentNodeOffset = entryNode;
+    if (entryNode == common::INVALID_OFFSET) {
+        return common::INVALID_OFFSET;
+    }
+    double lastMinDist = std::numeric_limits<float>::max();
+    const auto queryVector = info.embeddings->getEmbedding(node);
+    const auto currNodeVector = info.embeddings->getEmbedding(currentNodeOffset);
+    auto minDist = HNSWIndexUtils::computeDistance(info.distFunc, queryVector, currNodeVector,
+        info.embeddings->getDimension());
+    KU_ASSERT(lastMinDist >= 0);
+    KU_ASSERT(minDist >= 0);
+    while (minDist < lastMinDist) {
+        lastMinDist = minDist;
+        auto neighbors = graph->getNeighbors(transaction, currentNodeOffset);
+        for (const auto nbr : neighbors) {
+            const auto nbrVector = info.embeddings->getEmbedding(nbr);
+            const auto dist = HNSWIndexUtils::computeDistance(info.distFunc, queryVector, nbrVector,
+                info.embeddings->getDimension());
+            if (dist < minDist) {
+                minDist = dist;
+                currentNodeOffset = nbr;
+            }
+        }
+    }
+    return currentNodeOffset;
+}
+
+// NOLINTNEXTLINE(readability-make-member-function-const): Semantically non-const function.
+void InMemHNSWLayer::insertRel(transaction::Transaction* transaction, common::offset_t srcNode,
+    common::offset_t dstNode) {
+    const auto currentLen = graph->incrementCSRLength(srcNode);
+    if (currentLen >= info.maxDegree) {
```

Comment:
This maxDegree is degreeThresholdToShrink, right?

Reply:
Yep.